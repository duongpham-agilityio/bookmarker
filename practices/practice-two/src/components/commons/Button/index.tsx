import { ButtonHTMLAttributes } from 'react';

// Styles
import buttonStyles from 'components/commons/Button/index.module.css';

type Variant = 'default' | 'primary' | 'secondary' | 'danger';
type Size = 'small' | 'medium' | 'large';
type Border = 'b-md' | 'b-lg';
type Width = 'w-sm' | 'w-lg';
export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  label: string;
  leftIcon?: string;
  rightIcon?: string;
  variant?: Variant;
  border?: Border;
  size?: Size;
  width?: Width;
};

const ButtonIcon = ({
  icon,
  ...rest
}: {
  icon?: string;
  subTitle?: string;
}) => {
  return (
    <>
      {icon && (
        <img
          src={icon}
          alt="icon action"
          className={buttonStyles.icon}
          {...rest}
        />
      )}
    </>
  );
};

const Button = (props: ButtonProps) => {
  const {
    label,
    leftIcon,
    rightIcon,
    className,
    variant = 'default',
    size = 'medium',
    border = 'b-md',
    width = 'w-sm',
    type = 'button',
    ...rest
  } = props;

  const classes = `${buttonStyles.btn} ${buttonStyles[variant]}
  ${buttonStyles[size]} ${buttonStyles[border]}
  ${buttonStyles[width]} ${className ?? ''}`;

  return (
    <button className={classes} type={type} {...rest}>
      <ButtonIcon
        icon={leftIcon ?? ''}
        subTitle="This is the left icon for the button"
      />

      {label}

      <ButtonIcon
        icon={rightIcon ?? ''}
        subTitle="This is the right icon for the button"
      />
    </button>
  );
};

export default Button;
