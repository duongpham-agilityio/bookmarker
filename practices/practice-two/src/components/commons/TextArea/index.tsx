import { TextareaHTMLAttributes } from 'react';

// Styles
import styles from 'components/commons/TextArea/index.module.css';

export type TextAreaProps = TextareaHTMLAttributes<HTMLTextAreaElement>;

const TextArea = (props: TextAreaProps) => {
  const { value = '', ...rest } = props;

  return (
    <textarea className={styles.text} {...rest}>
      {value}
    </textarea>
  );
};

export default TextArea;
