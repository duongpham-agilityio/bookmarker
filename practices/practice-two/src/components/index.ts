export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Form, type FormProps } from './Form';
export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
