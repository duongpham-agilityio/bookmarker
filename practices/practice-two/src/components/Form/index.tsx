import { MouseEvent, memo } from 'react';

// Hooks
import { useForm } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';

// Types
import { Book } from 'types';

// Styles
import styles from 'components/Form/index.module.css';

// Assets
import UploadIcon from 'assets/icons/upload.svg';

export type FormProps = {
  value: Omit<Book, 'publishDate' | 'deletedAt' | 'createdAt' | 'updatedAt'> & {
    publishDate?: number;
  };
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: (_event: MouseEvent) => void;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit,
  } = useForm(data, type);

  return (
    <section
      className={styles.overlay}
      onClick={(e: MouseEvent) => {
        if (onClose) onClose(e);
      }}
    >
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmit}
        onClick={(e: MouseEvent) => {
          e.stopPropagation();

          resetRecommended([]);
        }}
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <ul className={styles.recommend}>
                  {booksRecommended.map((book, index) => {
                    return (
                      <li
                        className={styles.recommendItem}
                        key={index}
                        onClick={(event: MouseEvent) => {
                          event.preventDefault();

                          handleSelectRecommended(index);
                        }}
                      >
                        <p>{book.name}</p>
                      </li>
                    );
                  })}
                </ul>
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />
                {!imageName ? (
                  <>
                    <Button
                      label="Upload"
                      width="w-lg"
                      border="b-lg"
                      variant="primary"
                      leftIcon={UploadIcon}
                      onClick={() => {
                        refImage.current?.click();
                      }}
                    />
                  </>
                ) : (
                  <p
                    onClick={() => {
                      refImage.current?.click();
                    }}
                  >
                    {imageName}
                  </p>
                )}
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
            />
            <Button
              disabled={isUpload}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form);
