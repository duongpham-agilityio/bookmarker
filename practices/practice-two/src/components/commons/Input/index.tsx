import { InputHTMLAttributes, MouseEvent, Ref, forwardRef } from 'react';

// Style
import inputStyles from 'components/commons/Input/index.module.css';
import Button from '../Button';

export type InputProps = InputHTMLAttributes<HTMLInputElement> & {
  leftIcon?: string;
  rightIcon?: string;
  className?: string;
  leftIconClick?: (_event: MouseEvent) => void;
  rightIconClick?: (_event: MouseEvent) => void;
};

const InputIcon = ({
  icon,
  ...rest
}: {
  icon?: string;
  onClick?: (_event: MouseEvent) => void;
}) => {
  return (
    <Button
      label=""
      leftIcon={icon}
      style={{
        backgroundColor: 'transparent',
        padding: '0',
        width: 'fit-content',
        height: 'fit-content',
      }}
      {...rest}
    />
  );
};

const Input = (props: InputProps, ref: Ref<HTMLInputElement>) => {
  const {
    leftIcon = '',
    rightIcon = '',
    className = '',
    type = 'text',
    placeholder = 'Enter text...',
    leftIconClick,
    rightIconClick,
    ...rest
  } = props;

  return (
    <div className={`${inputStyles.input} ${className}`}>
      <InputIcon icon={leftIcon} onClick={leftIconClick} />

      <input
        className={inputStyles.value}
        type={type}
        placeholder={placeholder}
        ref={ref}
        {...rest}
      />

      <InputIcon icon={rightIcon} onClick={rightIconClick} />
    </div>
  );
};

export default forwardRef(Input);
