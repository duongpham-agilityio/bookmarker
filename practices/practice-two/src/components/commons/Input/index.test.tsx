import { fireEvent, render } from '@testing-library/react';

import { Input } from 'components/commons';

describe('Input', () => {
  it('match snapshot', () => {
    const { asFragment } = render(<Input />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', () => {
    const { getByPlaceholderText } = render(<Input placeholder="Enter" />);

    expect(getByPlaceholderText(/Enter/i)).toBeInTheDocument();
  });

  it('action', () => {
    const mockOnChange = jest.fn();
    const { getByPlaceholderText } = render(
      <Input placeholder="Enter" onChange={mockOnChange} />
    );
    const input = getByPlaceholderText(/Enter/i);

    fireEvent.change(input, {
      target: {
        value: '1',
      },
    });

    expect(mockOnChange).toHaveBeenCalled();
  });
});
