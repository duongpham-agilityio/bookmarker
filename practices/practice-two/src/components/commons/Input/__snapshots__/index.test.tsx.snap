// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input match snapshot 1`] = `
<DocumentFragment>
  <div
    class="undefined "
  >
    <input
      placeholder="Enter"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
